src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs

```diff
@@ -29,20 +28,20 @@ namespace MongoDB.Driver.Core.Servers
 {
     internal sealed class ServerMonitor : IServerMonitor
     {
-        private static readonly TimeSpan __minHeartbeatInterval = TimeSpan.FromMilliseconds(500);
-
-        private readonly ExponentiallyWeightedMovingAverage _averageRoundTripTimeCalculator = new ExponentiallyWeightedMovingAverage(0.2);
         private readonly ServerDescription _baseDescription;
-        private readonly CancellationTokenSource _cancellationTokenSource = new CancellationTokenSource();
         private volatile IConnection _connection;
         private readonly IConnectionFactory _connectionFactory;
+        private CancellationTokenSource _currentCheckCancellationTokenSource;
         private ServerDescription _currentDescription;
         private readonly EndPoint _endPoint;
+        private BuildInfoResult _handshakeBuildInfoResult;
         private HeartbeatDelay _heartbeatDelay;
-        private readonly TimeSpan _heartbeatInterval;
+        private readonly object _lock = new object();
+        private readonly CancellationTokenSource _monitorCancellationTokenSource;
```

Comment:
I don't think this needs to be a `CancellationTokenSource`. I think just a `CancellationToken` is appropriate. Also add comment:

    private readonly CancellationToken _monitorCancellationToken; // used to cancel the entire monitor

Reply:
we use CancellationTokenSource.Cancel() in `Dispose`. So I think, it's still required.

Reply:
That's unfortunate because it's APIs that allow for cancellation always take a Token and not a TokenSource.

Reply:
There's also an ownership problem here. Presumably the TokenSource is owned externally and passed to us in the constructor, so the following line of code in Dispose is not really correct (we are Disposing a resource we don't own):

    _monitorCancellationTokenSource.Dispose();

How about we take in a Token (not a TokenSource) in the constructor and create a LinkedTokenSource internally. The we own the TokenSource and have the right to Dispose it.

Reply:
Discussed offline. I removed cancellationToken and cancelationTokenSource from the ctor arguments. Now we will need to use `Dispose()`